src-electron/main-process/modules/wallet-rpc.js

```diff
@@ -1194,8 +1188,65 @@ export class WalletRPC {
     });
   }
 
-  transfer(password, amount, address, payment_id, priority, note, address_book = {}) {
-    crypto.pbkdf2(password, this.auth[2], 1000, 64, "sha512", (err, password_hash) => {
+  // submits the transaction to the blockchain, irreversible from here
+  async relayTransaction(metadataList, isBlink, addressSave, note) {
+    const { address, payment_id, address_book } = addressSave;
+    let failed = false;
+
+    // submit each transaction individually
+    let i;
+    for (i = 0; i < metadataList.length; i++) {
+      const hex = metadataList[i];
+      const params = {
+        hex,
+        isBlink
+      };
+      // don't try submit more txs if a prev one failed
+      if (failed) break;
+      await this.sendRPC("relay_tx", params)
```

Comment:
when using `async/await` you shouldn't be using `.then` or `.catch`.

Improvement:
```js
try {
  const data = await this.sendRPC("relay_tx", params);
  if (data.hasOwnProperty("error")) { ... }
  ...
} catch (e) {
   /// this.sendRPC threw an error
  failed = true
}
```

Reply:
without await, the `failed` flag is not set correctly since it will continue to run the down the function, before the `.then()` can execute, it wouldn't actually throw an error that could be caught either, rather it "succeeds" but with an `error` field.

Reply:
right that happens because the function is async.

These 2 pieces of code are identical:
```js
async do_something() {
  try {
    const something = await another();
    console.log(something);
  } catch (e) {
    console.error(e);
  }
}
```
```js
do_something() {
  another().then(console.log).catch(console.error);
}
```

When you start mixing them is when you run into weird issues like the one you mentioned (which occurs because this.sendRPC is actually a promise so you'd need to either await it or wrap all the logic into a `then`)

My suggestion would be to try and avoid the `.then` and `.catch` where possible and use `async/await` instead as it makes the code more clearer

Reply:
The failed functionality can be implemented like this:
```js
async function rpcCall() {
  let error;
  try {
    const data = await rpc.call();
    if data.hasOwnProperty("error") {
      const errMsg = data.error.message;
      error = errMsg.charAt(0).toUpperCase() + errMsg.slice(1);
      break;
    }
    ... // other logic
  } catch (e) {
    error = e.toString();
    break;
  }

  if (!error) {
    // Show success 
  } else {
    // Show error
  }
}
```